src/main/java/org/jboss/remoting3/ConnectionInfo.java

```diff
@@ -134,50 +147,47 @@ boolean connectionClosed(final AuthenticationConfiguration authenticationConfigu
         private final AuthenticationConfiguration authenticationConfiguration;
         private final IoFuture<Connection> attempt;
         private final Map<AuthenticationConfiguration, FutureResult<Connection>> pendingAttempts = new ConcurrentHashMap<>();
+        private final ReentrantReadWriteLock pendingAttemptsLock = new ReentrantReadWriteLock();
 
         MaybeShared(final AuthenticationConfiguration authenticationConfiguration, final IoFuture<Connection> attempt) {
             this.authenticationConfiguration = authenticationConfiguration;
             this.attempt = attempt;
         }
 
         IoFuture<Connection> getConnection(final EndpointImpl endpoint, final ConnectionKey key, final AuthenticationConfiguration authenticationConfiguration, boolean doConnect) {
-            if (authenticationConfiguration.equals(this.authenticationConfiguration)) {
-                return attempt;
-            } else {
-                FutureResult<Connection> futureResult = pendingAttempts.get(authenticationConfiguration);
-                if (futureResult != null) {
-                    return futureResult.getIoFuture();
+            try {
+                pendingAttemptsLock.readLock().lock();
```

Comment:
Two things to note:

* Lock acquisition should always happen before the ```try```, otherwise if the lock fails for whatever reason, the unlock is called when the thread doesn't own the lock, causing an IllegalMonitorStateException or similar.
* ReentrantReadWriteLock is too heavy for this usage since the inside of the block is short-lived and non-blocking.  It's probably better to just drop to a plain map and make it synchronized, and we can revisit later if there is substantial contention.